### PR TITLE
feat: NFTAccess: logic and tests for removeCollections

### PR DESCRIPTION
**Pull request by davidprepo**
Parent: https://github.com/prepo-io/prepo-monorepo/pull/583

### CODE REVIEW

**Comment by davidprepo** [comment]
Related: https://github.com/prepo-io/prepo-monorepo/pull/593

File: apps/smart-contracts/core/test/DepositHook.test.ts [hunk 1]
```diff
@@ -455,6 +455,80 @@ describe('=> DepositHook', () => {
...
+    it('is idempotent', async () => {
+      await depositHook.connect(deployer).setCollectionScores([firstERC721.address], [1])
+      expect(await depositHook.getCollectionScore(firstERC721.address)).to.eq(1)
+
+      await depositHook.connect(deployer).removeCollections([firstERC721.address])
+      expect(await depositHook.getCollectionScore(firstERC721.address)).to.eq(0)
+
+      await depositHook.connect(deployer).removeCollections([firstERC721.address])
+      expect(await depositHook.getCollectionScore(firstERC721.address)).to.eq(0)
+    })
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/test/DepositHook.test.ts:502]: Add idempotency test for removing multiple collections.

File: apps/smart-contracts/core/contracts/DepositHook.sol [hunk 2]
```diff
@@ -131,7 +131,15 @@ contract DepositHook is
     external
     override
     onlyRole(REMOVE_COLLECTIONS_ROLE)
-  {}
+  {
+    for (uint256 i = 0; i < _collections.length; i++) {
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/core/contracts/DepositHook.sol:135]: @davidprepo Did you verify in the `EnumerableMap` library if `remove` already returns early if a item already exists? So gas is not wasted deleting both enumeration indices and an entry.

**Review by ramenforbreakfast — APPROVED** [review]
LGTM besides one question and a nit.

File: apps/smart-contracts/core/contracts/DepositHook.sol (hunk 2 above)
Comment (XavierEkkel) [reply apps/smart-contracts/core/contracts/DepositHook.sol:135]: I just checked - it does skip the deletion logic if it doesn't exist in the collection https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/structs/EnumerableSet.sol#L87

so just calling `remove` is safe and also fine from a gas perspective

File: apps/smart-contracts/core/test/DepositHook.test.ts [hunk 3]
```diff
@@ -509,6 +509,113 @@ describe('=> DepositHook', () => {
...
+      expect(await depositHook.getCollectionScore(secondERC721.address)).to.eq(0)
+
+      await depositHook
+        .connect(deployer)
+        .removeCollections([firstERC721.address, secondERC721.address])
+      expect(await depositHook.getCollectionScore(firstERC721.address)).to.eq(0)
+      expect(await depositHook.getCollectionScore(secondERC721.address)).to.eq(0)
+    })
+
+    it('emits CollectionScoresChange when removing a single collection', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/DepositHook.test.ts:591]:
```suggestion
    it('emits CollectionScoresChange if single collection', async () => {
```

File: apps/smart-contracts/core/test/DepositHook.test.ts [hunk 4]
```diff
@@ -509,6 +509,113 @@ describe('=> DepositHook', () => {
...
+      expect(await depositHook.getCollectionScore(firstERC721.address)).to.eq(1)
+
+      const tx = await depositHook.connect(deployer).removeCollections([firstERC721.address])
+
+      await expect(tx)
+        .to.emit(depositHook, 'CollectionScoresChange')
+        .withArgs([firstERC721.address], [0])
+    })
+
+    it('emits CollectionScoresChange when removing multiple collections', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/DepositHook.test.ts:602]:
```suggestion
    it('emits CollectionScoresChange if multiple collections', async () => {
```

File: apps/smart-contracts/core/test/DepositHook.test.ts [hunk 5]
```diff
@@ -509,6 +509,113 @@ describe('=> DepositHook', () => {
...
+        .connect(deployer)
+        .removeCollections([firstERC721.address, secondERC721.address])
+      expect(await depositHook.getCollectionScore(firstERC721.address)).to.eq(0)
+      expect(await depositHook.getCollectionScore(secondERC721.address)).to.eq(0)
+
+      await depositHook
+        .connect(deployer)
+        .removeCollections([firstERC721.address, secondERC721.address])
+      expect(await depositHook.getCollectionScore(firstERC721.address)).to.eq(0)
+      expect(await depositHook.getCollectionScore(secondERC721.address)).to.eq(0)
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/DepositHook.test.ts:588]:
```suggestion
      await depositHook
        .connect(deployer)
        .removeCollections([firstERC721.address, secondERC721.address])
      expect(await depositHook.getCollectionScore(firstERC721.address)).to.eq(0)
      expect(await depositHook.getCollectionScore(secondERC721.address)).to.eq(0)

      await depositHook
        .connect(deployer)
        .removeCollections([firstERC721.address, secondERC721.address])

      expect(await depositHook.getCollectionScore(firstERC721.address)).to.eq(0)
      expect(await depositHook.getCollectionScore(secondERC721.address)).to.eq(0)
```

File: apps/smart-contracts/core/test/DepositHook.test.ts [hunk 6]
```diff
@@ -509,6 +509,113 @@ describe('=> DepositHook', () => {
...
+
+    it('is idempotent', async () => {
+      await depositHook.connect(deployer).setCollectionScores([firstERC721.address], [1])
+      expect(await depositHook.getCollectionScore(firstERC721.address)).to.eq(1)
+
+      await depositHook.connect(deployer).removeCollections([firstERC721.address])
+      expect(await depositHook.getCollectionScore(firstERC721.address)).to.eq(0)
+
+      await depositHook.connect(deployer).removeCollections([firstERC721.address])
+      expect(await depositHook.getCollectionScore(firstERC721.address)).to.eq(0)
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/DepositHook.test.ts:568]:
```suggestion
      await depositHook.connect(deployer).removeCollections([firstERC721.address])
      expect(await depositHook.getCollectionScore(firstERC721.address)).to.eq(0)

      await depositHook.connect(deployer).removeCollections([firstERC721.address])

      expect(await depositHook.getCollectionScore(firstERC721.address)).to.eq(0)
```

**Comment by ramenforbreakfast** [comment]
@davidprepo I have rebased this one as well, feel free to address comments and merge if everything is OK.

**Pull request merged**